File: protobuf-vs-json/to_protobuf.go
package main

import (
	"math/rand"
	"os"

	"github.com/golang/protobuf/proto"

	"benchmark/genproto"
)

func main() {
	vectorSize := 768
	vector := make([]float64, vectorSize)
	for i := 0; i < vectorSize; i++ {
		vector[i] = rand.Float64()
	}
	data, _ := proto.Marshal(&genproto.ProductVector{Id: 1, Vector: vector})
	os.WriteFile("/tmp/data.bin", data, 0644)
}

File: protobuf-vs-json/to_protobuf.py
import json
import random
import message_pb2

vector = [random.uniform(0, 1) for i in range(768)]
data = {"id": 1, "vector": vector}
msg = message_pb2.ProductVector(**data)

with open("/tmp/data.bin", "wb") as f:
    f.write(msg.SerializeToString())

data = {"vector": vector}
print(data)
vector = message_pb2.Vector(**data)
with open("/tmp/vector.bin", "wb") as f:
    f.write(vector.SerializeToString())
